### PR TITLE
WP list edit feature

### PR DESCRIPTION
**Pull request by furinvader**
### Description

This PR introduces inplace editing for work packages in the list view.
Supersedes #4153 

After trying `angular-xeditable` and fiddling around with the current inplace edit implementation, I've found that implementing a slim new module was the quickest solution.

Be aware, that currently the **list view does not work** as it's supposed to.
### TODO
- [x] Enable subject field editing
- [x] Enable drop down field editing (Type, Priority, ...)
- [x] Save the work package
- [x] @ulferts Use schema properties in edit form templates (linke `maxlength`, `minlength` and so on) [TODO](https://github.com/furinvader/openproject/blob/experimental/new_wp_table/1/frontend/app/components/wp-edit/wp-edit-field.directive.html#L8)
  - furinvader/openproject#8
- [x] @ulferts, @oliverguenther Automatically remove non writable properties from the work package, when using `save` [TODO](https://github.com/furinvader/openproject/blob/experimental/new_wp_table/1/frontend/app/components/api/api-work-packages/api-work-packages.config.ts#L73)
  controller
  - furinvader/openproject#7
- [x] @ulferts prevent calls to `work_packages/:id/:attribute_name` (e.g. http://localhost:3000/api/v3/work_packages/144/assignee) if the value is not set for a work package
  - furinvader/openproject#10
- [x] @ulferts Reenable the context menu on the work packages table
  - furinvader/openproject#9
  - furinvader/openproject#14
- [x] Fix karma, protractor, specs and cukes
  - furinvader/openproject#11
  - furinvader/openproject#12
- [x] Remove work package querying from experimental
  - [x] PaginationService, etc.
  - [x] Grouping of fields
  - https://github.com/furinvader/openproject/pull/16
- [x] Handle inserting / removing columns, since new columns are empty.
  - https://github.com/furinvader/openproject/pull/16
- [x] Add possibility for error handling to the `wp-edit.directive`
- [x] @ulferts remove empty rows from table
  - furinvader/openproject#13
- [x] `wp-td.directive.ts` -> Show values depending on their type as defined by the schema
  - furinvader/openproject#18
- [x] Display sums (grouping and total) according to the schema 
  - furinvader/openproject#18
- [x] assignee (and possibly responsible) is considered a text field when it should be a drop down
- [x] after save, the altered value is not displayed in the table (although the field is updated)
- [x] empty fields (e.g. assignee not set) can not be edited. This is due to the field having a height of `0px` when the value is not set. 
- [x] if the user lacks the permission to edit a work package (e.g. he is not logged in), there are errors in the console.
- [x] Attributes that are not required and that have a `select` as their input element should have a blank value ("-") as their first value.
  - [x] Delete attributes, e.g. assignee
- [x] The assignee attribute does not get it's current value selected when entering inline edit mode.
- [x] Modifying filters in saved queries will lead to an error in the experimental API.
- [x] After saving an inplace-edit once, subsequent edits will succeed, but no longer update the table cell.

`...`
#### Missing tests
- [x] `halTransform` service
- [ ] `apiWorkPackages` service
- [ ] `apiWorkPackages` config - test the added convenience methods

`...`
#### Miscellaneous / Ideas
- [ ] Update type definitions file - best done in development. This item should be checked when everything else is finished.
- ~~Make WP schema accessible (especially `allowedValues`) via the linked resource 'schema'. 
  Currently, we have to request the WP edit form first, which is quite unnecessary.~~
- ~~Make query filters work again. Implement the `apiMetaData` service, where the meta result of the experimental API is handled. There we will map the old properties to the new ones (column names, custom fields etc.)~~
#### Out of Scope
- Use the new `wpEdit` module in the details view, replacing `inplace-edit`.
- Update work package table after submit using the new values
- Enable two-way binding between list and details view
- Handle custom fields
- Classes of the [`wp-edit-field.module.ts`](https://github.com/furinvader/openproject/blob/experimental/new_wp_table/1/frontend/app/components/wp-edit/wp-edit-field.module.ts#L65-L86), that should be added, if they introduce any new behaviour.
- The updated on cell is not updated after editing a work package.
- Work packages groups are not updated after changing an attribute.
- Changing between list view, split view and fullscreen view.
- When clicking in a text field the cursor should be at the position where the user clicked.
- Misleading animation after changing the sorting criteria (there are some seconds waiting between the animation and the actual change)
##### Documentation
- [ ] Add documentation for the API services
- [ ] Add `jsdoc`/`ngdoc` and 'enforce' its usage
#### Addresses WPs
- https://community.openproject.org/work_packages/22816
- https://community.openproject.org/work_packages/22551/activity
- https://community.openproject.org/work_packages/22549/activity
- https://community.openproject.org/work_packages/22552/activity
- https://community.openproject.org/work_packages/22055/activity

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://furinvader-experimental-new-wp-table-1.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://furinvader-experimental-new-wp-table-1.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

**Comment by furinvader** [comment]
Before we start collaborating, please wait until the referenced pull requests are merged and the branch is rebased.

File: frontend/typings/restangular/restangular.d.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
+// Type definitions for Restangular v1.4.0
```
Comment (ulferts) [inline frontend/typings/restangular/restangular.d.ts:1]: The restangular version in the bower.json is 1.5.1. So either only this comment is outdated or the type definitions have to be updated as well.

Comment (furinvader) [inline frontend/typings/restangular/restangular.d.ts:1]: That is the content of a generated file, which we unfortunately have to commit.

**Comment by furinvader** [comment]
Sorry, I have rebased the branch and force pushed it :/

**Comment by furinvader** [comment]
After implementing the transformation service using actual Angular services that provide the necessary classes, I should do the same for the WP edit field classes.
The current implementation is pretty dirty with the `$injector` service being set as a static property for every field subclass.

**Comment by ulferts** [comment]
https://community.openproject.org/work_packages/22816

**Comment by ulferts** [comment]
* https://community.openproject.org/work_packages/22551/activity
* https://community.openproject.org/work_packages/22549/activity
* https://community.openproject.org/work_packages/22552/activity

**Comment by ulferts** [comment]
https://community.openproject.org/work_packages/22055/activity

**Pull request merged**